src/tools/ScenarioMeasurement/Startup/TraceEventParsers/PerfLabGenericEventSourceLTTngProviderParser.cs

```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    public event Action<OnMainTraceData> OnMain
+    {
+        add
+        {
+            source.RegisterEventTemplate(OnMainTemplate(value));
+        }
+        remove
+        {
+            source.UnregisterEventTemplate(value, PerfLabValues.OnMainEventId, ProviderGuid);
+        }
+    }
+
+    protected override string GetProviderName() => PerfLabValues.LTTngProviderName;
+
+    protected override IEnumerable<CtfEventMapping> EnumerateCtfEventMappings()
+    {
+        yield return new CtfEventMapping($"{PerfLabValues.LTTngProviderName}:{PerfLabValues.StartupEventName}", ProviderGuid, 0, PerfLabValues.StartupEventId, 0);
+        yield return new CtfEventMapping($"{PerfLabValues.LTTngProviderName}:{PerfLabValues.OnMainEventName}", ProviderGuid, 0, PerfLabValues.OnMainEventId, 0);
+    }
+
+    static volatile TraceEvent[] s_templates;
+    protected override void EnumerateTemplates(Func<string, string, EventFilterResponse> eventsToObserve, Action<TraceEvent> callback)
+    {
+        if (s_templates == null)
+        {
+            var templates = new TraceEvent[]
+            {
+                StartupTemplate(null),
+                OnMainTemplate(null),
+            };
+            s_templates = templates;
+        }
+        foreach (var template in s_templates)
+            if (eventsToObserve == null || eventsToObserve(template.ProviderName, template.EventName) == EventFilterResponse.AcceptEvent)
+                callback(template);
+    }
+
+    static StartupTraceData StartupTemplate(Action<StartupTraceData> action) => new StartupTraceData(action, PerfLabValues.StartupEventId, 1, PerfLabValues.StartupEventName, Guid.Empty, 0, string.Empty, ProviderGuid, PerfLabValues.LTTngProviderName);
+    static OnMainTraceData OnMainTemplate(Action<OnMainTraceData> action) => new OnMainTraceData(action, PerfLabValues.OnMainEventId, 1, PerfLabValues.OnMainEventName, Guid.Empty, 0, string.Empty, ProviderGuid, PerfLabValues.LTTngProviderName);
+
+    public abstract class PerfLabTraceData<TSelf> : TraceEvent
+        where TSelf : PerfLabTraceData<TSelf>
+    {
+        private event Action<TSelf> _target;
+
+        internal PerfLabTraceData(Action<TSelf> target, int eventID, int task, string taskName, Guid taskGuid, int opcode, string opcodeName, Guid providerGuid, string providerName)
+            : base(eventID, task, taskName, taskGuid, opcode, opcodeName, providerGuid, providerName)
+        {
+            _target = target;
+        }
+
+        protected override void Dispatch()
+        {
+            _target((TSelf)this);
+        }
+
+        protected override Delegate Target
+        {
+            get => _target;
+            set => _target = (Action<TSelf>)value;
+        }
+
+        public override string[] PayloadNames => Array.Empty<string>();
+
+        public override object PayloadValue(int index) => throw new NotSupportedException();
```

Comment:
Recommend that you return `null` here instead.  This will ensure that the behavior here matches what is expected, and since you don't have any event arguments, it should be safe to do so.

Reply:
Done.